### PR TITLE
fix: migrate remaining cl incentives

### PR DESCRIPTION
**Pull request by PaddyMc**
## What is the purpose of the change

This PR migrates the rest of the cl pools to the new scaling factor, this PR makes sure that we follow the same logic for the CL incentive logic for all pools

### How to test

To test this PR you need to use the `in-place-testnet` 

Download the latest snapshot and genesis

```bash
OSMOSIS_HOME=$HOME/.osmosisd

SNAPSHOT_URL=$(curl -sL https://snapshots.osmosis.zone/latest)
RPC_URL=https://rpc.osmosis.zone
ADDRBOOK_URL="https://rpc.osmosis.zone/addrbook"
GENESIS_URL=https://github.com/osmosis-labs/osmosis/raw/main/networks/osmosis-1/genesis.json

rm -rf $HOME/.osmosisd/

osmosisd init test --chain-id testing --home=$HOME/.osmosisd

# Copy genesis
echo -e "\nDownloading genesis file..."
wget $GENESIS_URL -O $OSMOSIS_HOME/config/genesis.json
echo ✅ Genesis file downloaded successfully.

# Download latest snapshot
echo -e "\nDownloading latest snapshot..."
wget -O - $SNAPSHOT_URL | lz4 -d | tar -C $OSMOSIS_HOME/ -xf -
echo -e ✅ Snapshot downloaded successfully.

# Run the node
osmosisd start --home=$HOME/.osmosisd
```

You need to install v23.0.8 to run the following command
```
osmosisd in-place-testnet edgenet osmo12smx2wdlyttvyzvzg54y2vnqwq2qjateuf7thj --trigger-testnet-upgrade v24
```

Wait until the node stops, build this branch and run
```
osmosisd start
```

This will run the v24 migration and migrate all remaining CL pool incentive accumlators

In the below are all the accumlators before and after the migration:

https://docs.google.com/spreadsheets/d/1CnhEKEjwOBKpsNYWrbwhpyfuS9Yva_qhhoCmJ6sbjJc/edit#gid=1264576604

How to test:

- Create an incentive for a pool
- Create a CL position in a tight range
- Swap tokens
- Check that rewards have been distributed

## Summary by CodeRabbit

- **New Features**
	- Introduced functionality to migrate remaining Concentrated Liquidity (CL) incentive accumulators.
- **Refactor**
	- Streamlined the calculation and claiming process for incentives in concentrated liquidity pools.
	- Adjusted scaling factors in incentive calculations to accommodate new parameters.
- **Bug Fixes**
	- Fixed issues related to the migration of accumulator scaling factors in concentrated liquidity pools.
	- Addressed truncation issues in core logic by scaling uptime accumulators.
- **Tests**
	- Updated and removed certain tests to reflect changes in incentive calculations and pool migrations.
- **Chores**
	- Removed outdated functionality and consolidated logic for better efficiency in handling incentives.

### CODE REVIEW

File: x/concentrated-liquidity/pool_test.go [hunk 1]
```diff
@@ -849,16 +846,14 @@ func (s *KeeperTestSuite) TestMigrateAccumulatorToScalingFactor() {
 	s.Require().NoError(err)
 
 	// Get claimable amount for position one before the migration
+	// NOTE: since the we enforce the scaling factor and never use the original scaling factor, this check is no longer applicable
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool_test.go:849]: Which check?

File: x/concentrated-liquidity/pool_test.go [hunk 2]
```diff
@@ -881,8 +876,31 @@ func (s *KeeperTestSuite) TestMigrateAccumulatorToScalingFactor() {
...
+	s.Require().NoError(err)
+
+	// Cross next right tick to update the tick accumulator by swapping
+	amtIn, _, _ = s.computeSwapAmounts(poolID, concentratedPool.GetCurrentSqrtPrice(), currentTick+100, false, false)
+	s.swapOneForZeroRight(poolID, sdk.NewCoin(USDC, amtIn.Ceil().TruncateInt()))
+
+	claimableIncentivesCompareOneAfterMigration, _, err := s.App.ConcentratedLiquidityKeeper.GetClaimableIncentives(s.Ctx, positionOneCompareID)
+
+	// Do the same swap as before the migration to get the same result
+	s.Require().Equal(claimableIncentivesCompareOneAfterMigration.String(), claimableIncentivesOneBeforeMigration.String())
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool_test.go:903]: This is a general repeat of what happens at the start of this test right? Could we use a helper func if so to improve readability?

File: x/concentrated-liquidity/types/migration.go [hunk 3]
```diff
@@ -8,3 +8,750 @@ var MigratedIncentiveAccumulatorPoolIDs = map[uint64]struct{}{
 	1292: {},
 	1431: {},
 }
+
+var FinalIncentiveAccumulatorPoolIDsToMigrated = map[uint64]struct{}{
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/migration.go:12]: Mentioned in another comment but I think this should live in the upgrade folder.

Comment (PaddyMc) [reply x/concentrated-liquidity/types/migration.go:12]: Fixed in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: x/concentrated-liquidity/pool_test.go (hunk 2 above)
Comment (PaddyMc) [reply x/concentrated-liquidity/pool_test.go:902]: Good idea but this actually took me a long time to refactor... and I couldn't get it to work as expected... :grimacing:

File: x/concentrated-liquidity/pool_test.go (hunk 1 above)
Comment (PaddyMc) [reply x/concentrated-liquidity/pool_test.go]: left over from previous iteration, removed in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: app/upgrades/v24/upgrades_test.go [hunk 4]
```diff
@@ -94,6 +96,65 @@ func (s *UpgradeTestSuite) TestUpgrade() {
...
+	s.Require().NoError(err)
+	s.Require().NotEmpty(migratedPoolBeforeUpgradeIncentives)
+
+	// Non-migrated pool claim
+	nonMigratedPoolBeforeUpgradeIncentives, _, err := s.App.ConcentratedLiquidityKeeper.GetClaimableIncentives(s.Ctx, lastPoolPositionID-1)
+	s.Require().NoError(err)
+	s.Require().NotEmpty(nonMigratedPoolBeforeUpgradeIncentives)
+
+	// Overwrite the migration list with the desired pool ID.
+	//oldMigrationList := concentratedtypes.MigratedIncentiveAccumulatorPoolIDs
```
[the remark this answers is not in the document]
Comment (PaddyMc) [reply app/upgrades/v24/upgrades_test.go]: must have left this in by accident, updated in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: app/upgrades/v24/upgrades_test.go [hunk 5]
```diff
@@ -94,6 +96,65 @@ func (s *UpgradeTestSuite) TestUpgrade() {
 	s.Require().Equal(twapRecord1, twapRecords[0])
 	s.Require().Equal(twapRecord2, twapRecords[1])
 
+	// INCENTIVES Setup
+	//
+
+	concentratedPoolIDs := []uint64{}
+
+	// Create two sets of all pools
```
[the remark this answers is not in the document]
Comment (PaddyMc) [reply app/upgrades/v24/upgrades_test.go]: good input, updated in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: app/upgrades/v24/upgrades.go [hunk 6]
```diff
@@ -47,9 +58,55 @@ func CreateUpgradeHandler(
...
 		keepers.ICAControllerKeeper.SetParams(ctx, icacontrollertypes.DefaultParams())
 
 		return migrations, nil
 	}
 }
+
+// migrateMainnetPools migrates the specified mainnet pools to the new accumulator scaling factor.
+func migrateMainnetPools(ctx sdk.Context, concentratedKeeper concentratedliquidity.Keeper) error {
+	poolIDsToMigrate := make([]uint64, 0, len(concentratedtypes.FinalIncentiveAccumulatorPoolIDsToMigrated))
+	for poolID := range concentratedtypes.FinalIncentiveAccumulatorPoolIDsToMigrated {
```
[the remark this answers is not in the document]
Comment (PaddyMc) [reply app/upgrades/v24/upgrades.go]: good input, updated in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: app/upgrades/v24/upgrades.go [hunk 7]
```diff
@@ -9,9 +11,18 @@ import (
...
 	"github.com/osmosis-labs/osmosis/v23/app/upgrades"
 
+	concentratedliquidity "github.com/osmosis-labs/osmosis/v23/x/concentrated-liquidity"
+	concentratedtypes "github.com/osmosis-labs/osmosis/v23/x/concentrated-liquidity/types"
 	incentivestypes "github.com/osmosis-labs/osmosis/v23/x/incentives/types"
 )
 
+const (
+	mainnetChainID = "osmosis-1"
+	// Edgenet is to function exactly the samas mainnet, and expected
```
[the remark this answers is not in the document]
Comment (PaddyMc) [reply app/upgrades/v24/upgrades.go]: ah sozo, updated in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: app/upgrades/v23/upgrades_test.go [hunk 8]
```diff
@@ -96,15 +96,17 @@ func (s *UpgradeTestSuite) TestUpgrade() {
 		s.App.BeginBlocker(s.Ctx, abci.RequestBeginBlock{})
 	})
 
+	// NOTE: these no longer work as expected in versions later than v23, in v23 we decide between the old and new scaling factor.
+	// Since we're removing this choice, these test are not worrking as expected
```
[the remark this answers is not in the document]
Comment (PaddyMc) [reply app/upgrades/v23/upgrades_test.go:100]: Nice catch, updated in 63ea98f2121e7fd321c348acc0bc65ccb5a69df3

File: x/concentrated-liquidity/types/migration.go (hunk 3 above)
Comment (p0mvn) [reply x/concentrated-liquidity/types/migration.go:12]: Do we want to remove this now that a copy exists in the upgrade folder?

**Review by p0mvn — COMMENTED** [review]
LGTM on a first pass. Great work!

I will run the tests from the PR description and do another pass tomorrow before the final approval

File: x/concentrated-liquidity/types/migration.go (hunk 3 above)
Comment (PaddyMc) [reply x/concentrated-liquidity/types/migration.go]: Moved in the previous commit, must have missed the delete. Ty :bow: 

Moved both of the pool id maps to their relevant upgrade folders => a410aa3c45d5808c8d602a48df951aeb5e6cfe24

**Comment by czarcas7ic** [comment]
@PaddyMc Do we still need the scalling factor as an input param here https://github.com/osmosis-labs/osmosis/blob/63ccc3668dffae20104d147923b993d5d8e82c67/x/concentrated-liquidity/incentives.go#L336-L349

**Comment by PaddyMc** [comment]
> @PaddyMc Do we still need the scale factor as an input param here
> 
> https://github.com/osmosis-labs/osmosis/blob/63ccc3668dffae20104d147923b993d5d8e82c67/x/concentrated-liquidity/incentives.go#L336-L349
> 
> Im fairly certain we use the same value every time

Yeah there's some clean up left to do after this PR specifically wrt scaling factor stuff, I think we can tackle that in subsequent PRs.

File: app/upgrades/v24/constants.go [hunk 9]
```diff
@@ -19,3 +19,750 @@ var Upgrade = upgrades.Upgrade{
 		Deleted: []string{},
 	},
 }
+
+var FinalIncentiveAccumulatorPoolIDsToMigrate = map[uint64]struct{}{
```
Comment (p0mvn) [inline app/upgrades/v24/constants.go:23]: What if new someone creates new CL pools before we release v24?

Would it make sense to make the system more dynamic? That is, migrate every CL pool other than the hard-coded white-listed list which is guaranteed to not change?

File: app/upgrades/v23/constants.go [hunk 10]
```diff
@@ -17,3 +17,12 @@ var Upgrade = upgrades.Upgrade{
...
 }
+
+var MigratedIncentiveAccumulatorPoolIDs = map[uint64]struct{}{
+	1423: {},
+	1213: {},
+	1298: {},
+	1297: {},
+	1292: {},
+	1431: {},
+}
```
Comment (p0mvn) [inline app/upgrades/v23/constants.go:28]: I see that these pools are marked green in the spreadsheet.

Why is pool 1323 green as well despite being migrated in v24?

File: app/upgrades/v24/constants.go (hunk 9 above)
Comment (czarcas7ic) [reply app/upgrades/v24/constants.go:23]: It wouldn't matter because after v23, all future pools created are on the new upscaled accumulator logic right?

Comment (PaddyMc) [reply app/upgrades/v24/constants.go:23]: I think it's fine, also based on the [Threshold](https://github.com/osmosis-labs/osmosis/pull/7805/files#diff-60e2a1beda87745518eb040a91479246d2073a318fd5b5d2f20aabf9755e7c07R125-R127) we will only ever migrate up to pool 1492

```
pool_id=1492
```

so all newly created pool are using the new scaling factor

Comment (PaddyMc) [reply app/upgrades/v24/constants.go:23]: But I've added all the latest CL pools to the migration, although they will be a NO-OP

File: app/upgrades/v23/constants.go (hunk 10 above)
Comment (PaddyMc) [reply app/upgrades/v23/constants.go:28]: Ah sorry my mistake, it will be migrated and I have highlighted it incorrectly :sweat_smile:

**Review by p0mvn — APPROVED** [review]
LGTM. Great work

**Pull request merged**